add read delays episode (#5)
epiverse-trace/tutorials-early: episodes/delays-reuse.Rmd

---
title: 'Reuse delay data'
teaching: 10
exercises: 2
editor_options: 
  chunk_output_type: inline
---

:::::::::::::::::::::::::::::::::::::: questions 

- How to get delay distributions from a systematic review?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

> - Get delays from a systematic review with `{epiparameter}`.

Author: Thank you for the specific suggestion. I'm interested to express then somewhere the line of "Such distributions will enable outbreak analysts to easily access these distributions for routine analysis". I'll look where it fits.

```suggestion
- Get delays from literature searches with `{epiparameter}`.
```

> - Get delays from a systematic review with `{epiparameter}`.

Author: replaced all terms in text in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/f6d1dbe3b4e2ae2f72d08ddb59bea07fabdab85d

- Get distribution parameters and summary statistics of delay distributions.

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: prereq

## Prerequisites

This episode requires you to be familiar with:

**Data science** : Basic programming with R.

> **Epidemic theory** : Epidemiological parameters. Time periods.

Author: Agree. For homogeneity across currently [online tutorials](https://epiverse-trace.github.io/tutorials-late/simulating-transmission.html#prerequisites), I'll use commas.

```suggestion
**Epidemic theory** : Epidemiological parameters, time periods.
```

:::::::::::::::::::::::::::::::::

## Introduction

> The [natural history](../learners/reference.md#naturalhistory) of an infectious disease shows that its development has a regularity from stage to stage. The time periods from an infectious disease inform about the timing of transmission and interventions.

Author: fixed in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/dd47ea823898638bdf9cb135f75b0a5871116546

![Definition of key time periods. From [Xiang et al, 2021](https://www.sciencedirect.com/science/article/pii/S2468042721000038)](fig/time-periods.jpg)

::::::::::::::::: callout

### Definitions

Look at the [glossary](../learners/reference.md) for the definitions of all the time periods of the figure above!

:::::::::::::::::::::::::

> However, early in an epidemic, modelling efforts can be delayed by the lack of a centralised resource that summarises input parameters for the disease of interest ([Nash et al., 2023](https://mrc-ide.github.io/epireview/)). Projects like `{epiparameter}` and `{epireview}` are building online catalogues following systematic review protocols that can help build models faster for coming outbreaks and epidemics from known pathogens and unknown ones related to known families of viruses.

Author: > It's fine to describe {epireview} as a systematic review as they are following a systematic protocol, however, same comment as above for {epiparameter} to rephrase systematic review.

Edited as "literature synthesis protocols" as a general reference to both in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/f6d1dbe3b4e2ae2f72d08ddb59bea07fabdab85d

Also used the proposed rephrase in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/87694a425302b5a14e52d0e7a9837db847ab6c04

<!-- Early models for COVID-19 used parameters from other coronaviruses. https://www.thelancet.com/article/S1473-3099(20)30144-4/fulltext -->

> To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to replace the `generation_time` input that we can use for `EpiNow2::epinow()`.

Author: Thanks for writing your main reason to use {epiparameter}. I emphasized the contrast between choosing within the analysis code against copy/pasting reported values on paper. Added in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/dcfc264d4cf5cbe644c97acfeaee6cd48bbfb920

> To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to replace the `generation_time` input that we can use for `EpiNow2::epinow()`.

Author: I opt not to show `epinow()` or `generation_times` yet in the first episode. We plan to show episodes in this order: `delays-reuse`, [`quantify-transmissibility`](https://epiverse-trace.github.io/tutorials-middle/quantify-transmissibility.html), `delays-functions`.

```r
epinow_estimates <- epinow(
  # cases
  reported_cases = example_confirmed[1:60],
  # delays
  generation_time = generation_time_opts(generation_time),
  # computation
  stan = stan_opts(
    cores = 4, samples = 1000, chains = 3,
    control = list(adapt_delta = 0.99)
  )
)
```

> To do this replacement, instead of plug-in numeric values to `EpiNow2::dist_spec()` to manually specify the **summary statistics** of the delay distribution, we are going to collect them from the library of epidemiological parameters provided by `{epiparameter}`:

Author: this now looks differently ([read the `md-output-PR`](https://github.com/epiverse-trace/tutorials-early/blob/md-outputs-PR-5/delays-reuse.md#the-problem)) and possibly with less ambiguity.

```r
generation_time <- dist_spec(
  mean = 3.6,
  sd = 3.1,
  max = 20,
  distribution = "lognormal"
)
```

Let's explore how we can access to time delays and get their **summary statistics** using `{epiparameter}`. We'll use the pipe `%>%` to connect some of their functions, so let's also call to the `{tidyverse}` package:

```{r,warning=FALSE,message=FALSE}
library(epiparameter)

> library(tidyverse)

Author: Added in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/ee7846c919376fb6b7ca4cd3d408d08f08d75f31

![Video from the MRC Centre for Global Infectious Disease Analysis, Ep 76. Science In Context - Epi Parameter Review Group with Dr Anne Cori (27-07-2023) at <https://youtu.be/VvpYHhFDIjI?si=XiUyjmSV1gKNdrrL>](fig/reproduction-generation-time.png)

In calculating the effective reproduction number ($R_{t}$), the *generation time* distribution is often approximated by the [serial interval](../learners/reference.md#serialinterval) distribution.
This frequent approximation is because it is easier to observe and measure the onset of symptoms than the onset of infectiousness.

![A schematic of the relationship of different time periods of transmission between an infector and an infectee in a transmission pair. Exposure window is defined as the time interval having viral exposure, and transmission window is defined as the time interval for onward transmission with respect to the infection time ([Chung Lau et al. 2021](https://academic.oup.com/jid/article/224/10/1664/6356465)).](fig/serial-interval-observed.jpeg)

> However, using the *serial interval* as an approximation of the *generation time* is primarily valid for diseases in which infectiousness starts after symptom onset ([Chung Lau et al. 2021](https://academic.oup.com/jid/article/224/10/1664/6356465)). In cases where infectiousness starts before symptom onset, the serial intervals can have negative values, which is the case of a pre-symptomatic transmission ([Nishiura et al. (2020)](https://www.ijidonline.com/article/S1201-9712(20)30119-3/fulltext#gr2)).

Author: Added in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/3baf6734702ba6800ccadc0c2ee1dfe41259b47d

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::: instructor

The objective of the assessment above is to assess the interpretation of a larger or shorter generation time.

::::::::::::::::::::::

## Extract epidemiological parameters

First, let's assume that the data set `example_confirmed` has COVID-19 observed cases. So, we need to find a reported generation time for COVID-19 or any other useful parameter for this aim. 

Let's start by looking at how many parameters we have in the epidemiological distributions database (`epidist_db`) for the `disease` named `covid`-19:

```{r}
epiparameter::epidist_db(
  disease = "covid"
)
```

From the `{epiparameter}` package, we can use the `epidist_db()` function to ask for any `disease` and also for a specific epidemiological distribution (`epi_dist`). 

Let's ask now how many parameters we have in the epidemiological distributions database (`epidist_db`) with the generation time using the string `generation`:

```{r}
epiparameter::epidist_db(
  epi_dist = "generation"
)

> ```

Reviewer: To answer one of your questions from the pull request description:

> Should we use partial words like "serial" instead of "serial interval" for epidist_db()?

I personally do not have a preference. {epiparameter} is setup to be flexible and accept both. If we are following the code standard of no partial matching then it might be best to specify in full as `generation time`. The downside of this is that is will not match database entries called `generation interval` or `generation period` so perhaps leaving as `generation` is best. I'll let you make the decision.

Author: I'll keep it with partial matching.

Currently, in the library of epidemiological parameters, we have one `generation` time entry for Influenza. Considering the abovementioned considerations, we can look at the `serial` intervals for `COVID`-19. Run this locally!

```{r,eval=FALSE}
epiparameter::epidist_db(
  disease = "COVID",
  epi_dist = "serial"
)
```

> With this query combination, we get more than one delay distribution. This output is an object class `<multi_epidist>`.

Reviewer: I'm not sure we need to mention the `<multi_epidist>` class. This is only in {epiparameter} to improve printing for long lists and does not have any functional purpose, so it may confuse learners more than help them.

Reviewer: Instead you can just describe it as a list of `<epidist>` objects.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/96d0ec8ce2c5487c8dae39a55d1bb9f43b84e754

Reviewer: I'm not sure the changes in 96d0ec8 are correct. If the output before was a `<multi_epidist>` this implies there was a list of `<epidist>` objects. It looks like the R code has not changed but the comments now state it is an `<epidist>`. Please check whether it is indeed an `<epidist>` or a list of `<epidist>` as this will matter for the user when working with these objects.

- Do you recognise the papers?

- Should it consider any other paper?

How does `{epiparameter}` do the collection and review of peer-reviewed literature? We invite you to read the vignette on ["Data Collation and Synthesis Protocol"](https://epiverse-trace.github.io/epiparameter/articles/data_protocol.html)!

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

## Select a single distribution

The `epiparameter::epidist_db()` function works as a filtering or subset function. Let's use the `author` argument to filter `Hiroshi Nishiura` parameters:

```{r}
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial",
  author = "Hiroshi"
) %>%
  epiparameter::list_distributions()
```

We still get more than one epidemiological parameter. We can set the `single_epidist` argument to `TRUE` to only one:

```{r}
epiparameter::epidist_db(
  disease = "covid",
  epi_dist = "serial",
  author = "Hiroshi",
  single_epidist = TRUE
)
```

::::::::::::::::: callout

### How does `single_epidist` works?

Looking at the help documentation for `?epiparameter::epidist_db()`:

- If multiple entries match the arguments supplied and `single_epidist = TRUE`,

> - Then, the parameterised`⁠<epidist>`⁠ with the largest sample size will be returned.

Reviewer: I think these should be the same bullet point.

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/a33b7d3553b767a569e6b306e3a3d553debfd787

- If multiple entries are equal after this sorting, the first entry will be returned.

> What does a parametrised `<epidist>` is? Look at `?is_parameterised`.

Reviewer: This question needs rewording. What about "What is a parameterised `<epidist>`?"

Author: added in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/47148b9d6b88c8ad1b924aae101f1b17b8250037

A severity parameter like the duration of hospitalisation could add to the information needed about the bed capacity in response to an outbreak ([Cori et al., 2017](https://royalsocietypublishing.org/doi/10.1098/rstb.2016.0371)).

For Ebola: 

- What is the reported point estimate and uncertainty of the mean duration of health care and case isolation?

::::::::::::::::: hint

An informative delay should measure the time from symptom onset to recovery or death.

::::::::::::::::::::::

::::::::::::::::: solution

```{r,eval=TRUE}
# one way to get the list of all the available parameters
epidist_db(disease = "all") %>%
  list_distributions() %>%
  as_tibble() %>%
  distinct(epi_distribution)

ebola_severity <- epidist_db(
  disease = "ebola",
  epi_dist = "onset to discharge"
)

# point estimate
ebola_severity$summary_stats$mean
# 95% confidence intervals
ebola_severity$summary_stats$mean_ci
# limits of the confidence intervals
ebola_severity$summary_stats$mean_ci_limits
```

::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::

<!--
## Concept map

update it from last epiparameter test 
-->

::::::::::::::::::::::::::::::::::::: keypoints 

> - Use `{epiparameter}` to access the systematic review catalogue of epidemiological delay distributions.

Reviewer: Same comment as above about systematic reviews.

Author: resolved in https://github.com/epiverse-trace/tutorials-early/pull/5/commits/f6d1dbe3b4e2ae2f72d08ddb59bea07fabdab85d

![Definition of key time periods. From [Xiang et al, 2021](https://www.sciencedirect.com/science/article/pii/S2468042721000038)](fig/time-periods.jpg)

::::::::::::::::: callout

### Definitions

Look at the [glossary](../learners/reference.md) for the definitions of all the time periods of the figure above!

:::::::::::::::::::::::::

However, early in an epidemic, modelling efforts can be delayed by the lack of a centralised resource that summarises input parameters for the disease of interest ([Nash et al., 2023](https://mrc-ide.github.io/epireview/)). Projects like `{epiparameter}` and `{epireview}` are building online catalogues following literature synthesis protocols that can help parametrise models by easily accessing a comprenhensive library of previously estimated epidemiological parameters from past outbreaks.

<!-- Early models for COVID-19 used parameters from other coronaviruses. https://www.thelancet.com/article/S1473-3099(20)30144-4/fulltext -->

> To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to *choose* one specific set of epidemiological parameters from the literature, instead of copy/pasting them *by hand*, to plug them in to an `{EpiNow2}` analysis workflow.

Author: ```suggestion
To exemplify how to use `{epiparameter}` in your analysis pipeline, our goal in this episode will be to *choose* one specific set of epidemiological parameters from the literature, instead of *copying-and-pasting* them by hand, to plug them into an `{EpiNow2}` analysis workflow.
```

<!-- In this episode, we'll learn how to choose one specific set of epidemiological parameters from the literature and then get their **summary statistics** using `{epiparameter}`.  -->

> Let's start loading the `{epiparameter}` package. We'll use the pipe `%>%` to connect some of their functions, so let's also call to the `{magrittr}` package:

Author: Getting error in challenge solution for `dplyr::distinct()`. Also using `tibble::as_tibble()` here, so better moving this back to {tidyverse}

```suggestion
Let's start loading the `{epiparameter}` package. We'll use the pipe `%>%` to connect some of their functions, some `{tibble}` and `{dplyr}` functions, so let's also call to the `{tidyverse}` package:
```

```{r,warning=FALSE,message=FALSE}
library(epiparameter)

> library(magrittr)

Author: ```suggestion
library(tidyverse)
```
